servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java

```diff
@@ -175,6 +206,65 @@ public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> for
+     *
+     * @param host host to connect to, resolved by default using a DNS {@link ServiceDiscoverer} for every new
+     * connection. This will also be used for the {@link HttpHeaderNames#HOST} together with the {@code port}. Use
+     * {@link SingleAddressHttpClientBuilder#unresolvedAddressToHost(Function)} if you want to override that value
+     * or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @param port port to connect to
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
+            final String host, final int port) {
+        return forSingleAddressResolveOnDemand(HostAndPort.of(host, port));
+    }
+
+    /**
+     * Creates a {@link SingleAddressHttpClientBuilder} for an address with default {@link LoadBalancer} and DNS {@link
+     * ServiceDiscoverer} that will run resolutions before creating a new connection.
+     * <p>
+     * Important side effects to take into account:
+     * <ol>
+     *     <li>The total latency for opening a new connection will be increased by a latency of DNS resolution.</li>
+     *     <li>If the target host has more than one resolved address, created clients loose ability to load balance for
+     *     each request. Instead, the balancing will happen only for every new created connection.</li>
+     *     <li>Created clients won't be able to move/shift traffic based on changes in DNS records until the remote
+     *     server closes existing connections.</li>
+     * </ol>
+     * <p>
+     * The returned builder can be customized using {@link SingleAddressHttpClientBuilderProvider}.
+     *
+     * @param address the {@code UnresolvedAddress} to connect to, resolved using a DNS {@link ServiceDiscoverer} for
+     * every new connection. This address will also be used for the {@link HttpHeaderNames#HOST} header value.
+     * Use {@link SingleAddressHttpClientBuilder#unresolvedAddressToHost(Function)} if you want to override that
+     * value or {@link SingleAddressHttpClientBuilder#hostHeaderFallback(boolean)} if you want to disable this behavior.
+     * @return new builder for the address
+     * @see SingleAddressHttpClientBuilderProvider
+     */
+    public static SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> forSingleAddressResolveOnDemand(
+            final HostAndPort address) {
+        return applyProviders(address,
+                new DefaultSingleAddressHttpClientBuilder<>(address, unresolvedServiceDiscoverer()))
+                // Apply after providers to let them see these customizations.
+                .appendConnectionFactoryFilter(ResolvingConnectionFactoryFilter.INSTANCE)
+                .retryServiceDiscoveryErrors(NoRetriesStrategy.INSTANCE);
```

Comment:
@daschl for context: this approach requires 3 (or at least 2) settings on the builder:
1. swapping default SD with `unresolvedServiceDiscoverer()`;
2. adding `ResolvingConnectionFactoryFilter`
3. (optional) disabling `retryServiceDiscoveryErrors` bcz retries are not required for `unresolvedServiceDiscoverer()`

I was trying to find a way how to simplify users' experience for configuring all required components with a single API call. As the result I introduced these new factories on `HttpClients`. 
I also considered something like `HttpClientFunctions` that will effectively be `UnaryOperator<SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress>>` and users will apply it to the builder when the want this mode, but found it a bit difficult for users to discover. 

I'm open to any other ideas how to make users experience better.

Reply:
I agree that wrapping this in a user-friendly way is good - see my additional comments on the potential other overload.